exchange/legacy.go

```diff
@@ -0,0 +1,356 @@
+		IsDebug: isDebug,
+		App:     req.App,
+		Device:  req.Device,
+		// PBSUser is excluded because rubicon is the only adapter which reads from it, and they're supporting OpenRTB directly
+		// SDK is excluded because that information doesn't exist in OpenRTB.
+		// Bidders is excluded because no legacy adapters read from it
+		User:   req.User,
+		Cookie: cookie,
+		Url:    url,
+		Domain: domain,
+		// Start is excluded because no legacy adapters read from it
+	}, nil
+}
+
+func toAccountId(req *openrtb.BidRequest) (string, error) {
+	if req.Site != nil && req.Site.Publisher != nil {
+		return req.Site.Publisher.ID, nil
+	}
+	if req.App != nil && req.App.Publisher != nil {
+		return req.App.Publisher.ID, nil
+	}
+	return "", errors.New("bidrequest.site.publisher.id or bidrequest.app.publisher.id required for legacy bidders.")
+}
+
+func toTransactionId(req *openrtb.BidRequest) (string, error) {
+	if req.Source != nil {
+		return req.Source.TID, nil
+	}
+	return "", errors.New("bidrequest.source.tid required for legacy bidders.")
+}
+
+func toSecure(req *openrtb.BidRequest) (secure int8, err error) {
+	secure = -1
+	for _, imp := range req.Imp {
+		if imp.Secure != nil {
+			thisVal := *imp.Secure
+			if thisVal == 0 {
+				if secure == 1 {
+					err = errors.New("bidrequest.imp[i].secure must be consistent for legacy bidders. Mixing 0 and 1 are not allowed.")
+					return
+				}
+				secure = 0
+			} else if thisVal == 1 {
+				if secure == 0 {
+					err = errors.New("bidrequest.imp[i].secure must be consistent for legacy bidders. Mixing 0 and 1 are not allowed.")
+					return
+				}
+				secure = 1
+			}
+		}
+	}
+	if secure == -1 {
+		secure = 0
+	}
+
+	return
+}
+
+func toLegacyBidder(req *openrtb.BidRequest, name openrtb_ext.BidderName) (*pbs.PBSBidder, []error) {
+	adUnits, errs := toPBSAdUnits(req)
+	if len(adUnits) > 0 {
+		return &pbs.PBSBidder{
+			BidderCode: string(name),
+			// AdUnitCode is excluded because no legacy adapters read from it
+			// ResponseTime is excluded because no legacy adapters read from it
+			// NumBids is excluded because no legacy adapters read from it
+			// Error is excluded because no legacy adapters read from it
+			// NoCookie is excluded because no legacy adapters read from it
+			// NoBid is excluded because no legacy adapters read from it
+			// UsersyncInfo is excluded because no legacy adapters read from it
+			// Debug is excluded because legacy adapters only use it in nil-safe ways.
+			//   They *do* write to it, though, so it may be read when unpacking the response.
+			AdUnits: adUnits,
+		}, errs
+	} else {
+		return nil, errs
+	}
+}
+
+func toPBSAdUnits(req *openrtb.BidRequest) ([]pbs.PBSAdUnit, []error) {
+	adUnits := make([]pbs.PBSAdUnit, 0, len(req.Imp))
+	var errs []error = nil
+	for i := 0; i < len(req.Imp); i++ {
+		adUnit, err := toPBSAdUnit(&(req.Imp[i]))
+		if err != nil {
+			errs = append(errs, err)
+		} else {
+			adUnits = append(adUnits, *adUnit)
+
```

Comment:
I notice you are likely copying a new adUnit struct into the adUnits slice. This increases the overhead of this adapter. Higher performance could be achieved by filling the slice with adUnits, passing a pointer to each one to toPBSAdUnit(), and the cutting off a number at the end corresponding to the number of errors. Probably not worth the effort at this stage, since this is just a stop gap until all adapters can support OpenRTB natively.

Reply:
that makes sense. Normally I don't bother optimizing stuff, but... this was so quick that I might as well.